src/main/java/spireMapOverhaul/zones/ambushed/AmbushedZone.java

```diff
@@ -0,0 +1,384 @@
+        HashSet<String> selectedCardIDs = new HashSet<>();
+
+        for (AbstractCard card : cards) {
+            AbstractCard newCard = getCardForShop(card.rarity, card.type, isColored, selectedCardIDs);
+            if (newCard != null) {
+                newCards.add(newCard);
+            }
+        }
+
+        cards.clear();
+        cards.addAll(newCards);
+    }
+
+    private AbstractCard getCardForShop(AbstractCard.CardRarity rarity, AbstractCard.CardType type, boolean isColored, HashSet<String> selectedCardIDs) {
+        List<String> filteredCardIDs = getFilteredCardIDsForShop(rarity, type, isColored, selectedCardIDs);
+
+        // Fallback 1: If no cards of the specified type, try another type
+        if (filteredCardIDs.isEmpty()) {
+            for (AbstractCard.CardType fallbackType : AbstractCard.CardType.values()) {
+                if (fallbackType != type) {
+                    filteredCardIDs = getFilteredCardIDsForShop(rarity, fallbackType, isColored, selectedCardIDs);
+                    if (!filteredCardIDs.isEmpty()) break;
+                }
+            }
+        }
+
+        // Fallback 2: If still no cards, try a different rarity
+        if (filteredCardIDs.isEmpty()) {
+            for (AbstractCard.CardRarity fallbackRarity : AbstractCard.CardRarity.values()) {
+                if (fallbackRarity != rarity) {
+                    filteredCardIDs = getFilteredCardIDsForShop(fallbackRarity, type, isColored, selectedCardIDs);
+                    if (!filteredCardIDs.isEmpty()) break;
+                }
+            }
+        }
+
+        if (filteredCardIDs.isEmpty()) {
+            return null;
+        }
+
+        String selectedCardID = filteredCardIDs.get(AbstractDungeon.cardRng.random(filteredCardIDs.size() - 1));
+        selectedCardIDs.add(selectedCardID);
+
+        return CardLibrary.getCard(selectedCardID).makeCopy();
+    }
+
+
+    private List<String> getFilteredCardIDsForShop(AbstractCard.CardRarity rarity, AbstractCard.CardType type, boolean isColored, HashSet<String> selectedCardIDs) {
+        // Filter your card draw cards by rarity, type, and whether it's colored or colorless
+        return cardDrawCardIDs.stream()
+                .filter(id -> isCardForShop(id, rarity, type, isColored))
+                .filter(id -> !selectedCardIDs.contains(id)) // Exclude already selected card IDs
+                .collect(Collectors.toList());
+    }
+
+
+    private boolean isCardForShop(String cardID, AbstractCard.CardRarity rarity, AbstractCard.CardType type, boolean isColored) {
+        AbstractCard card = CardLibrary.getCard(cardID);
+        boolean isCardColored = card.color != AbstractCard.CardColor.COLORLESS;
+        return card != null && card.rarity == rarity && card.type == type && isCardColored == isColored;
+    }
+
+    @Override
+    public void postCreateShopPotions(ShopScreen screen, ArrayList<StorePotion> potions) {
```

Comment:
I think this is not a good idea. Maybe just replace one potion with a draw potion. This seems overkill, and yoU're unlikely to want this many anyway. Also too much repetition in the potion choices

Reply:
I would make the argument that due to the amount of zones there are and how small my zone is in terms of height + width, the odds of someone naturally wandering into a shop that happens to fall in my zone are very slim.

Also, replacing just one potion with a draw potion would be too petty to add as a modification I think. I am satisfied with the current setup, although maybe I could make it so one potion of each rarity is guaranteed to appear? So the player has the option to purchase a Swift Potion, a Gambler's Brew, and a Smoke Bomb (lol) / Snecko Oil. I do think three Swift Potions in one shop feels pretty bad.

Reply:
Yeah, I'd be okay with that.